PWGHF/HFC/TableProducer/correlatorDMesonPairs.cxx

```diff
@@ -0,0 +1,702 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file correlatorDMesonPairs.cxx
+/// \brief D0(bar) and DPlus(Minus) correlator task - data-like, MC-reco and MC-kine analyses.
+///
+/// \author Fabio Colamaria <fabio.colamaria@ba.infn.it>, INFN Bari
+/// \author Andrea Tavira García <tavira-garcia@ijclab.in2p3.fr>, IJCLab Orsay
+
+#include "Common/Core/TrackSelection.h"
+#include "Common/DataModel/TrackSelectionTables.h"
+#include "Framework/AnalysisTask.h"
+#include "Framework/HistogramRegistry.h"
+#include "Framework/runDataProcessing.h"
+#include "PWGHF/DataModel/CandidateReconstructionTables.h"
+#include "PWGHF/DataModel/CandidateSelectionTables.h"
+
+using namespace o2;
+using namespace o2::framework;
+using namespace o2::framework::expressions;
+using namespace o2::aod::hf_cand;
+using namespace o2::aod::hf_cand_2prong;
+using namespace o2::aod::hf_cand_3prong;
+using namespace o2::constants::math;
+
+///
+/// Returns deltaPhi value in range [-pi/2., 3.*pi/2], typically used for correlation studies
+///
+double getDeltaPhi(double phiD, double phiDbar)
+{
+  return RecoDecay::constrainAngle(phiDbar - phiD, -o2::constants::math::PIHalf);
+}
+
+// histogram binning definition
+const int massAxisBins = 120;
+const double massAxisMin = 1.5848;
+const double massAxisMax = 2.1848;
+const int phiAxisBins = 32;
+const double phiAxisMin = 0.;
+const double phiAxisMax = 2. * o2::constants::math::PI;
+const int yAxisBins = 100;
+const double yAxisMin = -5.;
+const double yAxisMax = 5.;
+const int ptDAxisBins = 180;
+const double ptDAxisMin = 0.;
+const double ptDAxisMax = 36.;
+
+using MCParticlesPlus2Prong = soa::Join<aod::McParticles, aod::HfCand2ProngMcGen>;
+using MCParticlesPlus3Prong = soa::Join<aod::McParticles, aod::HfCand3ProngMcGen>;
+
+struct HfCorrelatorDMesonPairs {
+  SliceCache cache;
+  Preslice<aod::HfCand2Prong> perCol2Prong = aod::hf_cand::collisionId;
+  Preslice<aod::HfCand3Prong> perCol3Prong = aod::hf_cand::collisionId;
+
+  Produces<aod::D0Pair> entryD0Pair;
+  Produces<aod::D0PairRecoInfo> entryD0PairRecoInfo;
+  Produces<aod::DPlusPair> entryDPlusPair;
+  Produces<aod::DPlusPairRecoInfo> entryDPlusPairRecoInfo;
+
+  Configurable<bool> verbose{"verbose", false, "Enable debug mode"};
+  Configurable<int> particlePdgCode{"particlePdgCode", 421, "PDG code of particle to analyse"};
```

Comment:
It is needed for the kinematic cuts, because for D0's I have the condition "candidate.hfflag() & 1 << o2::aod::hf_cand_2prong::DecayType::D0ToPiK", and for D+'s I have "candidate.hfflag() & 1 << o2::aod::hf_cand_3prong::DecayType::DplusToPiKPi".

At first I also needed it to do the analysis at Gen level, but I could change it so that it doesn't need to use particlePdgCode.

Reply:
Also, changing the Gen level code, I realised that I was filling the D0Pair table for both D0's and D+'s, so I have fixed that.